### PR TITLE
Validate secret and set status on IngressRoute if secret is invalid 

### LINKED ISSUE #517: Validation: Set status on IngressRoute if TLS secret invalid
If an IngressRoute specifies a secret used for TLS certificates and that secret does not exist, the `status` field on the IR should be updated with an appropriate message. 

Current implementation ignores the secret if issue: https://github.com/heptio/contour/blob/master/internal/dag/dag.go#L375

// ref https://github.com/heptio/contour/issues/458

### PR DESCRIPTION
**Pull request by stevesloka**
Fixes #517 by setting the status of an IngressRoute if the TLS secret is not found or malformed. 

Example:
```
Status:
  Current Status:  invalid
  Description:     TLS Secret not found or is malformed
```
 
Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: internal/dag/builder.go [hunk 1]
```diff
@@ -215,7 +213,7 @@ func (b *builder) lookupSecret(m meta, validate func(*v1.Secret) bool) *Secret {
 		b.secrets = make(map[meta]*Secret)
 	}
 	b.secrets[s.toMeta()] = s
-	return s
+	return s, false
```
Comment (davecheney) [inline internal/dag/builder.go]: Is this correct? if we return the secret then it _is_ found.

File: internal/dag/builder.go [hunk 2]
```diff
@@ -515,15 +513,21 @@ func (b *builder) computeIngressRoutes() {
 		if tls := ir.Spec.VirtualHost.TLS; tls != nil {
 			// attach secrets to TLS enabled vhosts
 			m := splitSecret(tls.SecretName, ir.Namespace)
-			if sec := b.lookupSecret(m, validSecret); sec != nil && b.delegationPermitted(m, ir.Namespace) {
+			sec, secretInvalidOrNotFound := b.lookupSecret(m, validSecret)
+			if sec != nil && b.delegationPermitted(m, ir.Namespace) {
```
Comment (davecheney) [inline internal/dag/builder.go]: This is the only place that the second boolean parameter is begin used. Would this workinstead

```go
sec := b.lookupSecret(m, validSecret)
secretInvalidOrNotFound := sec == nil
```

Comment (stevesloka) [reply internal/dag/builder.go]: Yup that works! For some reason, I was making that more complicated than it needed to be.

File: internal/dag/builder.go [hunk 3]
```diff
@@ -197,15 +197,13 @@ func (b *builder) addTCPService(svc *v1.Service, port *v1.ServicePort, strategy
...
 func (b *builder) lookupSecret(m meta, validate func(*v1.Secret) bool) *Secret {
 	if s, ok := b.secrets[m]; ok {
 		return s
 	}
 	sec, ok := b.source.secrets[m]
-	if !ok {
-		return nil
-	}
-	if !validate(sec) {
+	if !ok || !validate(sec) {
```
Comment (davecheney) [inline internal/dag/builder.go]: I'm not sure this is an improvement in readability, i had to squint real hard to convince myself that if ok is true then the logic falls through to validate(sec)

Given that both validate functions, validSecret and validCA will work just fine, ie not panic, but return false if passed a nil. You could rewrite this as

```go
sec := b.source.secrets[m]
if !validate(sec) {
       return nil
}
```

Comment (davecheney) [reply internal/dag/builder.go]: opps, my bad, this wasn't correct. if sec == nil then both validate functions will panic. My recommendation is to roll back the change to this function. thanks for the comment,

Comment (stevesloka) [reply internal/dag/builder.go]: I've revered this back to the way it was

File: internal/dag/builder.go [hunk 4]
```diff
@@ -197,6 +197,7 @@ func (b *builder) addTCPService(svc *v1.Service, port *v1.ServicePort, strategy
 	return s
 }
 
+// lookupSecret returns a kuberntes secret if found or nil if not
```
Comment (davecheney) [inline internal/dag/builder.go]: s/kuberntes/kubernetes/

**Pull request merged**